src/backend/distributed/connection/connection_management.c

```diff
@@ -1095,21 +1080,67 @@ AfterXactHostConnectionHandling(ConnectionHashEntry *entry, bool isCommit)
 		}
 		else
 		{
-			/* reset per-transaction state */
-			ResetRemoteTransaction(connection);
-			ResetShardPlacementAssociation(connection);
-
-			/* reset copy state */
-			connection->copyBytesWrittenSinceLastFlush = 0;
-
-			UnclaimConnection(connection);
+			/*
+			 * Preserve healthy session lifespan connections.
+			 */
+			PreserveConnection(connection);
 
 			cachedConnectionCount++;
 		}
 	}
 }
 
 
+/*
+ * ShouldShutdownConnection returns true if either one of the followings is true:
+ * - The connection is citus initiated.
+ * - Current cached connections is already at MaxCachedConnectionPerWorker
+ * - Connection is forced to close at the end of transaction
+ * - Connection is not in OK state
+ * - Connection is not idle.
+ */
+static bool
+ShouldShutdownConnection(MultiConnection *connection, const int cachedConnectionCount)
+{
+	bool isCitusInitiatedBackend = false;
+
+	/*
+	 * When we are in a backend that was created to serve an internal connection
+	 * from the coordinator or another worker, we disable connection caching to avoid
+	 * escalating the number of cached connections. We can recognize such backends
+	 * from their application name.
+	 */
+	if (application_name != NULL && strcmp(application_name, CITUS_APPLICATION_NAME) == 0)
+	{
+		isCitusInitiatedBackend = true;
+	}
+
+	return isCitusInitiatedBackend ||
+		   cachedConnectionCount >= MaxCachedConnectionsPerWorker ||
+		   connection->forceCloseAtTransactionEnd ||
+		   PQstatus(connection->pgConn) != CONNECTION_OK ||
+		   !RemoteTransactionIdle(connection);
+}
+
+
+/*
+ * PreserveConnection preserves the given connection for later usage by
+ * resetting its states.
+ */
+static void
+PreserveConnection(MultiConnection *connection)
```

Comment:
If I look at what the function does it actually resets the connection to its initial state. Maybe `ResetConnection`?